src/libprojectM/PCM.cpp

```diff
@@ -24,381 +24,528 @@
+void PCM::getSpectrum(float *data, int channel, size_t samples, float smoothing)
+{
+    assert(channel == 0 || channel == 1);
+    _updateFFT();
 
-   start = (start + samples) % maxsamples;
+    float *spectrum = channel == 0 ? spectrumL : spectrumR;
+    if (smoothing == 0)
+    {
+        size_t count = samples <= FFT_LENGTH ? samples : FFT_LENGTH;
+        for (size_t i = 0; i < count; i++)
+            data[i] = spectrum[i];
+        for (size_t i = count; i < samples; i++)
+            data[0] = 0;
+    }
+    else
+    {
+        float l2 = 0, l1 =0 , c = 0, r1, r2;
+        r1 = spectrum[0]; r2 = spectrum[0+1];
+        for (size_t i = 0; i < samples; i++)
+        {
+            l2 = l1;
+            l1 = c;
+            c = r1;
+            r1 = r2;
+            r2 = (i + 2) >= samples ? 0 : spectrum[i + 2];
+            data[i] = (l2 + 4 * l1 + 6 * c + 4 * r1 + r2) / 16.0;
+        }
+    }
+}
 
-   newsamples+=samples;
-   if (newsamples>maxsamples) newsamples=maxsamples;
-     numsamples = getPCMnew(pcmdataR,1,0,waveSmoothing,0,0);
-    getPCMnew(pcmdataL,0,0,waveSmoothing,0,1);
-    getPCM(vdataL,FFT_LENGTH,0,1,0,0);
-    getPCM(vdataR,FFT_LENGTH,1,1,0,0);
+void PCM::_updateFFT()
+{
+    if (newsamples > 0)
+    {
+        _updateFFT(0);
+        _updateFFT(1);
+        newsamples = 0;
+    }
 }
 
+void PCM::_updateFFT(size_t channel)
+{
+    assert(channel == 0 || channel == 1);
+
+    double *freq = channel==0 ? freqL : freqR;
+    _copyPCM(freq, channel, FFT_LENGTH*2);
+    rdft(FFT_LENGTH*2, 1, freq, ip, w);
+
+    // compute magnitude data (m^2 actually)
+    float *spectrum = channel==0 ? spectrumL : spectrumR;
+    for (size_t i=1 ; i<FFT_LENGTH ; i++)
+    {
+        double m2 = (freq[i * 2] * freq[i * 2] + freq[i * 2 + 1] * freq[i * 2 + 1]);
+        spectrum[i-1] = m2 * ((double)i)/FFT_LENGTH;
+    }
+    spectrum[FFT_LENGTH-1] = freq[1] * freq[1];
+}
 
-void PCM::addPCM16(short PCMdata[2][512])
+inline double constrain(double a, double mn, double mx)
 {
-  int i,j;
-  int samples=512;
-
-	 for(i=0;i<samples;i++)
-	   {
-	     j=i+start;
-         if ( PCMdata[0][i] != 0 && PCMdata[1][i] != 0 ) {
-	         PCMd[0][j%maxsamples]=(PCMdata[0][i]/16384.0);
-	         PCMd[1][j%maxsamples]=(PCMdata[1][i]/16384.0);
-          } else {
-             PCMd[0][j % maxsamples] = (float)0;
-             PCMd[1][j % maxsamples] = (float)0;
-          }
-	   }
-
-	 // printf("Added %d samples %d %d %f\n",samples,start,(start+samples)%maxsamples,PCM[0][start+10]);
-
- start+=samples;
- start=start%maxsamples;
-
- newsamples+=samples;
- if (newsamples>maxsamples) newsamples=maxsamples;
-
-    numsamples = getPCMnew(pcmdataR,1,0,waveSmoothing,0,0);
-    getPCMnew(pcmdataL,0,0,waveSmoothing,0,1);
-    getPCM(vdataL,FFT_LENGTH,0,1,0,0);
-    getPCM(vdataR,FFT_LENGTH,1,1,0,0);
+    return a>mx ? mx : a<mn ? mn : a;
 }
 
+// pull data from circular buffer
+void PCM::_copyPCM(float *to, int channel, size_t count)
```

Comment:
I think it'd be good to keep any buffers we have as small as possible to not lag between sound and visual
What about using something like memcpy here for efficiency?

Reply:
If you're concerned about how the data is being provided to PCM.  That can be an issue, I have strange behavior with spotify on linux, but not with  browser streaming for instance.  I don't know what that is all about.

Once we have the data, I've never been able to show that any of the audio handling is really a problem.  The data is just so small.